### PR TITLE
Setup virtool_workflow_runtime

### PR DESCRIPTION
**Pull request by BlakeASmith**

### CODE REVIEW

File: virtool_workflow/context.py [hunk 1]
```diff
@@ -23,7 +23,8 @@ class Context:
     def __init__(
             self,
             on_update: Optional[UpdateListener] = None,
-            on_state_change: Optional[StateListener] = None
+            on_state_change: Optional[StateListener] = None,
+            event_loop: asyncio.AbstractEventLoop = asyncio.get_event_loop()
```
Comment (igboyes) [inline virtool_workflow/context.py:27]: Don't need to explicitly pass event loop. Just use `asyncio.get_event_loop()` wherever necessary.

File: virtool_workflow/execute.py [hunk 2]
```diff
@@ -1,48 +1,67 @@
...
             await on_each(wf, ctx)
-        update = await __run_step(step, wf, ctx, on_error)
-        if update:
-            await ctx.send_update(update)
+        update = await _run_step(step, wf, ctx, on_error)
+        await ctx.send_update(update)
             
             
-async def __inc_step(_, ctx):
+async def _inc_step(wf, ctx):
```
Comment (igboyes) [inline virtool_workflow/execute.py:62]: `async` not required as this is currently defined.

File: virtool_workflow_runtime/db.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        self.db_connection_url = getenv(DATABASE_CONNECTION_URL_ENV,
+                                           default=DATABASE_CONNECTION_URL_DEFAULT)
+        self.client = AsyncIOMotorClient(self.db_connection_url)
+        self.db = DB(self.client, enqueue_change=self._on_update)
+        self.jobs = self.db.jobs
+
+        self._init_job_entry()
+
+
+    def _init_job_entry(self):
```
Comment (igboyes) [inline virtool_workflow_runtime/db.py:30]: Done on server during API request handling.

**Review by igboyes — CHANGES_REQUESTED** [review]
I really like where this is going. Minor changes or comments.

Having the separate folders `virtool_workflow` and `virtool_workflow_runtime` does not follow conventional Python package layout. Reason for this? I would think they could be merged.

**Comment by BlakeASmith** [comment]
> I really like where this is going. Minor changes or comments.
> 
> Having the separate folders `virtool_workflow` and `virtool_workflow_runtime` does not follow conventional Python package layout. Reason for this? I would think they could be merged.

`virtool_workflow_runtime` contains code relating to the `workflow` CLI tool and the runtime specific code (such as sending updates to the `jobs` database). It's a tool for running workflows and creating workflow containers, not a library. A developer could still import `virtool_workflow_runtime` if they really wanted too, but the package is only intended for use by the CLI tool. 

`virtool_workflow` on the other hand is the library for building workflows. Having the package structure like this keeps the runtime code, which is not intended to be used inside workflows, "behind the curtain" so to speak. When a developer imports `virtool_workflow` they won't get any runtime specific items. 

I was inspired to set things up this way from looking at the source for pytest https://github.com/pytest-dev/pytest/tree/master/src 

The source is separated into two packages. `_pytest` contains all of the pytest application code and `pytest` 
contains the outward facing library code (what you get from `import pytest`).

**Comment by igboyes** [comment]
Thanks for the well documented response.

**Pull request merged**